populus/plugin.py

```diff
@@ -12,17 +24,135 @@
+)
 
 CACHE_KEY_MTIME = "populus/project/compiled_contracts_mtime"
 CACHE_KEY_CONTRACTS = "populus/project/compiled_contracts"
 
+BASE_DIR = os.path.dirname(__file__)
+ASSETS_DIR = os.path.join(BASE_DIR, 'assets')
+
+
+testing_project_dir = None
+testing_context = None
+testing_user_config_path = None
+#
+# Hooks
+#
+
+
+def get_populus_option(cmdline_option, ini_option, environ_var, early_config, args):
+
+    for arg in args:
+        option, _, val = arg.partition("=")
+        if option == cmdline_option:
+            return val
+    else:
+        if early_config.getini(ini_option):
+            return early_config.getini(ini_option)
+        else:
+            return os.environ.get(environ_var)
+
+
+def pytest_addoption(parser):
+
+    parser.addoption("--populus-user-config", help="populus global user config json file")
+    parser.addini("populus_user_config", "populus global user config json file")
+    parser.addoption("--populus-project", help="populus project root directory")
+    parser.addini("populus_project", "populus project root directory")
+    parser.addoption("--populus-context", help="populus testing context")
+    parser.addini("populus_context", "populus testing context")
+
+
+def pytest_load_initial_conftests(early_config, parser, args):
+
+    global testing_user_config_path
+    testing_user_config_path = get_populus_option(
+        cmdline_option="--populus-user-config",
+        ini_option="populus_user_config",
+        environ_var="POPULUS_USER_CONFIG",
+        early_config=early_config,
+        args=args
+    )
+    if not testing_user_config_path:
+        testing_user_config_path = get_user_default_json_config_file_path()
+
+    global testing_context
+    testing_context = get_populus_option(
+        cmdline_option="--populus-context",
+        ini_option="populus_context",
+        environ_var="POPULUS_PYTEST_CONTEXT",
+        early_config=early_config,
+        args=args
+    )
+
+    if not testing_context:
+        testing_context = 'user'
+
+    global testing_project_dir
+    testing_project_dir = get_populus_option(
+        cmdline_option="--populus-project",
+        ini_option="populus_project",
+        environ_var="POPULUS_PYTEST_PROJECT",
+        early_config=early_config,
+        args=args
+    )
+
+    if testing_project_dir:
+        testing_project_dir = os.path.abspath(testing_project_dir)
+        if not os.path.exists(get_json_config_file_path(testing_project_dir)):
+            raise FileNotFoundError(
+                "No populus project found for testing in {testing_project_dir}".format(
+                    testing_project_dir=testing_project_dir
+                )
+            )
+    else:
+        testing_project_dir = str(parser.extra_info['rootdir'])
+        try:
+            if args[0][0] != "-":
+                testing_project_dir = os.path.abspath(args[0])
+        except:
+            pass
+
 
 @pytest.fixture()
 def project(request):
+
+    global testing_project_dir
```

Comment:
I'm not seeing anything that makes it clear why we need globals.  why can't these just be fixtures?

Reply:
the conftest injection will not work

Reply:
See https://github.com/Aviah/populus/pull/12